### PR TITLE
Some tweaks in Typers

### PR DESCRIPTION
**Pull request by joroKr21**
* Use `foreach` instead of `map` to `completeInfo`
  * Remove unused pattern binders
  * Add `tailrec` annotations
  * `List()` -> `Nil`
  * `filterNot` -> `filter`
  * Remove overzealous parentheses
  * And some other tweaks

Let me know if stylistic changes are too much,
I may have gotten carried away 😄

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -3491,10 +3489,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           vparams.lengthIs <= MaxFunctionArity &&
           (formal baseType FunctionClass(vparams.length) match {
             case TypeRef(_, _, formalargs) =>
-              ( exists2(formalargs, vparams)((formal, vparam) =>
-                        vparam.tpt.isEmpty && (tparams exists formal.contains))
-                && (tparams forall isLowerBounded)
-              )
+              exists2(formalargs, vparams) { (formal, vparam) =>
+                vparam.tpt.isEmpty && (tparams exists formal.contains)
```
Comment (diesalbla) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I understand we want to get rid of infix notation for methods. 

```suggestion
                vparam.tpt.isEmpty && tparams.exists(formal.contains)
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -4212,23 +4210,21 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
      */
     def packSymbols(hidden: List[Symbol], tp: Type): Type = global.packSymbols(hidden, tp, context.owner)
 
-    def isReferencedFrom(ctx: Context, sym: Symbol): Boolean = (
-       ctx.owner.isTerm && (ctx.scope.exists { dcl => dcl.isInitialized && (dcl.info contains sym) }) || {
+    def isReferencedFrom(ctx: Context, sym: Symbol): Boolean =
+       ctx.owner.isTerm && ctx.scope.exists(dcl => dcl.isInitialized && (dcl.info contains sym)) || {
```
Comment (diesalbla) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]:
```suggestion
       ctx.owner.isTerm && ctx.scope.exists(dcl => dcl.isInitialized && dcl.info.contains(sym)) || {
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -4160,7 +4158,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           val info = annInfo(typedAnn)
           // check message and since args to deprecated
           // when deprecated gets more params, those other args may also be defaults.
-          def oneTwoPunch(p: Tree => Boolean) = (0 to 1).exists(info.argAtIndex(_).map(p).getOrElse(false))
+          def oneTwoPunch(p: Tree => Boolean) = (0 to 1).exists(info.argAtIndex(_).exists(p))
```
Comment (diesalbla) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]:
```suggestion
          def oneTwoPunch(p: Tree => Boolean) = 
            info.argAtIndex(0).exists(p)) || info.argAtIndex(1).exists(p)
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 4]
```diff
@@ -1971,10 +1971,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     }
 
     private def ensurePredefParentsAreInSameSourceFile(template: Template) = {
-      val parentSyms = template.parents map (_.symbol) filterNot (_ == AnyRefClass)
+      val parentSyms = template.parents.collect { case parent if parent.symbol != AnyClass => parent.symbol }
```
Comment (diesalbla) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: We can merge this `collect` with the `exists` below.
```suggestion
      val hasUndefinedParent = template.parents.exists { parent => 
        parent.symbol != AnyClass  && parent.symbol.associatedFile != PrefedModuleFile
      }
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 1 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Yeah there are many more of those around but I can change the places I'm touching 👍

**Review by diesalbla — APPROVED** [review]

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 5]
```diff
@@ -2087,7 +2086,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       if (!clazz.isTrait && clazz.isNonBottomSubClass(ConstantAnnotationClass)) {
         val (pConstrOpt, auxConstrs) = body3.filter(s => s.isInstanceOf[DefDef] && s.symbol.isConstructor).splitAt(1)
         for (p <- pConstrOpt) p.symbol.paramss match {
-          case List(ps) =>
+          case List(_) =>
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: This is just a weird way of saying `if (p.symbol.paramss.length != 1) CANSA(p, clazz)`. The snippet also suffers from what Diego always points out, creating lists which are discarded, instead of just iterating. There is always a primary ctor.
```
val ctors = body3.iterator.filter(s => s.isInstanceOf[DefDef] && s.symbol.isConstructor)
val p = ctors.next()
if (p.symbol.paramss.length != 1) ConstantAnnotationNeedsSingleArgumentList(p, clazz)
for (c <- ctors) AuxConstrInConstantAnnotation(c, clazz) 
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 6]
```diff
@@ -141,10 +142,11 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     def privates[T <: Tree](typer: Typer, owner: Symbol, tree: T): T =
       if (owner.isJavaDefined) tree else check(typer, owner, EmptyScope, WildcardType, tree)
 
+    @tailrec
     private def check[T <: Tree](typer: Typer, owner: Symbol, scope: Scope, pt: Type, tree: T): T = {
       this.owner = owner
       this.scope = scope
-      hiddenSymbols = List()
+      hiddenSymbols = Nil
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:148]: OK but scalac does this for you. It seems to me there are previous PRs about this, but I don't remember which is preferred.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 7]
```diff
@@ -1152,10 +1154,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           }
           @inline def warnNumericWiden(tpSym: Symbol, ptSym: Symbol): Unit =
             if (!isPastTyper) {
-              val isInharmonic = (
+              val isInharmonic =
                 tpSym == IntClass  && ptSym == FloatClass ||
                 tpSym == LongClass && (ptSym == FloatClass || ptSym == DoubleClass)
-              )
+
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Someone started the parens style to make sure expressions aren't accidentally split: add a blank line and it changes. I don't feel a need for parens around if-else. Here I'm more on the fence. I'll be more comfortable with doti-style indentation verification.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 8]
```diff
@@ -1408,12 +1410,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
-          ( silent   (_.typedArgs(args.map(_.duplicate), mode))
-              filter (xs => !(xs exists (_.isErrorTyped)))
-                 map (xs => adaptToArguments(qual, name, xs, WildcardType, reportAmbiguous, saveErrors))
-              orElse ( _ => reportError)
-          )
-        case _            =>
+          silent(_.typedArgs(args.map(_.duplicate), mode))
+            .filter(!_.exists(_.isErrorTyped))
+            .fold(reportError)(adaptToArguments(qual, name, _, WildcardType, reportAmbiguous, saveErrors))
+        case _ =>
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1415]: A striking improvement! My favorite are stale aligned arrows like this one!

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 5 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Yeah I thought about refactoring this but I couldn't 100% convince myself that there is always a primary constructor.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 6 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:148]: I find it more readable but maybe others disagree

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 9]
```diff
@@ -1531,7 +1531,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
                 /* check all base classes, since derived value classes might lurk in refinement parents */
                 if (acc.tpe.typeSymbol.baseClasses exists (_.isDerivedValueClass))
                   context.error(acc.pos, "value class may not wrap another user-defined value class")
-                checkEphemeral(clazz, body filterNot (stat => stat.symbol != null && stat.symbol.accessedOrSelf == paramAccessor))
+                checkEphemeral(clazz, body.filter(stat => stat.symbol == null || stat.symbol.accessedOrSelf != paramAccessor))
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: `filterNot` doesn't delegate to `filter`, both call `filterCommon`. In this case, it's easier to read the "guard clause" `stat.symbol != null` as a short-circuit. If you name the condition `def referencesUnderlying` then it's very readable `body.filterNot(referencesUnderlying)`. Approximately,
```
def referencesUnderlying(sym: Symbol) = sym != null && sym.accessedOrSelf == paramAccessor
decls.toList.find(x => x.isMethod && referencesUnderlying(x)) match {
  case Some(accessor) =>
    if (accessor.tpe.typeSymbol.baseClasses.exists(_.isDerivedValueClass))
      context.error(accessor.pos, "value class may not wrap another user-defined value class")
    checkEphemeral(clazz, body.filterNot(stat => referencesUnderlying(stat.symbol)))
}
```

Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I can do that. I found the double negation very confusing although it's quite common in my native tongue.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 10]
```diff
@@ -1971,10 +1971,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     }
 
     private def ensurePredefParentsAreInSameSourceFile(template: Template) = {
-      val parentSyms = template.parents map (_.symbol) filterNot (_ == AnyRefClass)
       val PredefModuleFile = PredefModule.associatedFile
-      if (parentSyms exists (_.associatedFile != PredefModuleFile))
-        context.error(template.pos, s"All parents of Predef must be defined in ${PredefModuleFile}.")
+      if (template.parents.exists(p => p.symbol != AnyClass && p.symbol.associatedFile != PredefModuleFile))
+        context.error(template.pos, s"All parents of Predef must be defined in $PredefModuleFile.")
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1976]: Noting switch to AnyClass.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 11]
```diff
@@ -3335,9 +3333,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       // 'accessor' and 'accessed' are so similar it becomes very difficult to
       //follow the logic, so I renamed one to something distinct.
       def accesses(looker: Symbol, accessed: Symbol) = accessed.isLocalToThis && (
-        (accessed.isParamAccessor)
-          || (looker.hasAccessorFlag && !accessed.hasAccessorFlag && accessed.isPrivate)
-        )
+        accessed.isParamAccessor || (looker.hasAccessorFlag && !accessed.hasAccessorFlag && accessed.isPrivate)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Parens also extra.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 12]
```diff
@@ -3491,10 +3488,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           vparams.lengthIs <= MaxFunctionArity &&
           (formal baseType FunctionClass(vparams.length) match {
             case TypeRef(_, _, formalargs) =>
-              ( exists2(formalargs, vparams)((formal, vparam) =>
-                        vparam.tpt.isEmpty && (tparams exists formal.contains))
-                && (tparams forall isLowerBounded)
-              )
+              exists2(formalargs, vparams) { (formal, vparam) =>
+                vparam.tpt.isEmpty && tparams.exists(formal.contains)
+              } && tparams.forall(isLowerBounded)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Without trying to understand it, maybe:
```
case (formal, Function(vparams, _)) if vparams.exists(_.tpt.isEmpty) && vparams.lengthIs <= MaxFunctionArity && tparams.forall(isLowerBounded) =>
  formal.baseType(FunctionClass(vparams.length)) match {
    case TypeRef(_, _, formalargs) => exists2(formalargs, vparams)((formal, vparam) => vparam.tpt.isEmpty && tparams.exists(formal.contains))
    case _ => false
  }
case _ => false
```

**Comment by som-snytt** [comment]
Somewhere in the contributers guide, they should make it clear that you can't go too far!

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 6 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:148]: I don't disagree.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 10 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:1976]: Surely you mean `AnyRefClass`

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 12 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Now is it better or worse to first get the length of `vparams`?

Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Or it just doesn't matter...

Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Oh `tparams` can be extracted even further up

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I'm resisting understanding the code, but I bet it doesn't matter. :)

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 10 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:1976]: I think it was `filterNot(_ == AnyRefClass)` and now `exists(_.symbol != AnyClass)`. I'm not sure, without waking up completely, if it matters.

Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:1976]: It was `map(_.symbol).filterNot(_ == AnyRefClass)` 🤷‍♂️

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 12 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I bet the condition `<= MaxFunctionArity` is true most of the time so better to get the length first. Splitting hairs here 😄

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 13]
```diff
@@ -2085,12 +2085,10 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
-        val (pConstrOpt, auxConstrs) = body3.filter(s => s.isInstanceOf[DefDef] && s.symbol.isConstructor).splitAt(1)
-        for (p <- pConstrOpt) p.symbol.paramss match {
-          case List(ps) =>
-          case _ => ConstantAnnotationNeedsSingleArgumentList(p, clazz)
-        }
-        for (c <- auxConstrs) AuxConstrInConstantAnnotation(c, clazz)
+        val ctors = body3.iterator.collect { case method: DefDef if method.symbol.isConstructor => method }
+        val primary = ctors.next() // there is always a primary constructor
+        if (primary.symbol.paramss.lengthIs != 1) ConstantAnnotationNeedsSingleArgumentList(primary, clazz)
+        ctors.foreach(AuxConstrInConstantAnnotation(_, clazz))
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:2090]: Nice! My only footnote (since I learned it this year) is that `isInstanceOf` doesn't checkout outer pointers, but pattern will if it can. Probably it doesn't matter here.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 14]
```diff
@@ -2366,7 +2365,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         if (!isPastTyper && meth.isPrimaryConstructor) {
           for (vparams <- ddef.vparamss; vd <- vparams) {
             if (vd.mods.isParamAccessor) {
-              vd.symbol setAnnotations (vd.symbol.annotations filter AnnotationInfo.mkFilter(ParamTargetClass, defaultRetention = true))
+              val sym = vd.symbol
+              sym.setAnnotations(sym.annotations.filter(AnnotationInfo.mkFilter(ParamTargetClass, defaultRetention = true)))
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:2368]: Sacrificing your red vs green metric for the sake of your naming principles! I can respect that.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 15]
```diff
@@ -3485,16 +3482,14 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
-              ( exists2(formalargs, vparams)((formal, vparam) =>
-                        vparam.tpt.isEmpty && (tparams exists formal.contains))
-                && (tparams forall isLowerBounded)
-              )
+              exists2(formalargs, vparams) { (formal, vparam) =>
+                vparam.tpt.isEmpty && tparams.exists(formal.contains)
+              }
             case _ =>
               false
           })
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3494]: I was hoping to lose the extra paren. :)  More seriously, hoisting the lower bounds check is probably cheap but it's incurred for every apply with undet tparams. It short-circuits looking at value args and other checks. A win!

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 16]
```diff
@@ -4160,7 +4156,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           val info = annInfo(typedAnn)
           // check message and since args to deprecated
           // when deprecated gets more params, those other args may also be defaults.
-          def oneTwoPunch(p: Tree => Boolean) = (0 to 1).exists(info.argAtIndex(_).map(p).getOrElse(false))
+          def oneTwoPunch(p: Tree => Boolean) = info.argAtIndex(0).exists(p) || info.argAtIndex(1).exists(p)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: It slightly dilutes the pun. :) The previous comment was expecting they'd add `forRemoval` to deprecated; it wasn't necessary to hardcode the expected count. Also FTR the `blockUsesDefault` test that follows is hard to follow.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 17]
```diff
@@ -4498,7 +4492,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           case _                              => false
         }
       }
-      acceptsApplyDynamicWithType(qual, name) map { tp =>
+      acceptsApplyDynamicWithType(qual, name).map { _ =>
         // If tp == NoType, pass only explicit type arguments to applyXXX.  Not used at all
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: `tp` is mentioned in the obscure comment

**Review by som-snytt — APPROVED** [review]
Could have gone further! :)

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 17 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Huh, we have special accommodation for scala-virtualized? It's kinda dead.
Also how could all of this be used when it's private?
I think `acceptsApplyDynamicWithType` should just return a boolean as the name implies.

Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Also this is an instance of "the invisible gorilla"

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 13 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:2090]: TIL

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 16 above)
Comment (joroKr21) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: You mean it dilutes the punch? 😄

**Review by lrytz — APPROVED** [review]
👍

**Comment by SethTisue** [comment]
is any of this likely to affect performance, or is it only stylistic? wasn't sure whether to put the "performance" label on

**Comment by som-snytt** [comment]
It was about 50/50 hair-splitting and side-splitting.

Definitely helps developer performance. Hopefully without sacrificing runtime performance. Or forcing any infos. Which sounds like the latest euphemism for "coercive interrogation."

So I would omit the performance label, just in case there is a regression. Why attract attention?
